source/_posts/2017-11-22-Challenges-when-making-Today-Widgets.md

```diff
@@ -0,0 +1,260 @@
+---
+title:  Challenges when making Today Widgets
+date: 2017-22-21 10:10:57
+tags: ios,widgets,today,extension
+authorIds:
+- dori
+categories:
+- iOS
+---
+
+Apple introduced the Today extension in iOS 9, more commonly called widgets.
+
+Apple describes widgets like so:
+
+Widgets give users quick access to information that’s important right now. For example, users open the Today view to check current stock prices or weather conditions, see today’s schedule, or perform a quick task such as marking an item as done. Users tend to open the Today view frequently, and they expect the information they’re interested in to be instantly available.
+
+## Setup
+
+Setting up a Today extension in your iOS app is fairly straight forward. In your xCode project click File->New->Target and select today extension.
+
+Make sure your widget bundle id is not the same as your apps bundle id. So if your apps bundle is “com.yourcompany.yourappname” the widgets bundle id could/should look like this “com.yourcompany.yourappname.widget”
+
+This will create a new target in your app and a new folder. The content of this folder should be familiar with any iOS developer. It contains
+
+1. TodayExtensionViewController.swift
+2. Storyboard file
+3. Plist
+4. Asset catalogue
+
+I like to rename the files from TodayExtension to Widget as that makes more sense to me.
+
+## Widget Creation
+
+Creating the UI for a widget should also be very simple for any developer that has used storyboards and auto-layout before.
+Just add the UI elements you wish to use to the storyboard and create the necessary outlets in the ViewController.
+
+This is what the storyboard could look like.
+
+<p align=center>
+<img src=https://cdn-laravel.vapor.cloud/image/nstack/translate_values/ScreenShot2017-11-22at104935_5ahcPXPuRk.png?width=280 style=";margin:20px;">
+
+## NSWidgetProviding
```

Comment:
Just a minor personal preference thing, but consider wrapping these code references. E.g. NSWidgetProviding vs `NSWidgetProviding`. It sometimes makes it more clear that we're referring to specific types.